### DISCUSSION TITLE
UP load test using multi-ue

### DISCUSSION THREAD

**Discussion by abousselmi** [HewlettPackard/PacketRusher#91] [Q&A]
Hello @linouxis9 

Were you able to send traffic using more than 1 UE ?

If I create 2 UEs with `multi-ue`, for some reason yet unknown to me (gtp5g kernel module limitation ?) only 1 single UE VRF can be used to send traffic (more specifically the first one that is used). I was able to reach roughly 5 Gbps using iperf3.

In this configuration, we should use a separate gNB per UE,

I tried to use TAP instead of VRF but it seems there is a problem with the CLI, it doesn't take into account the `--tunnel-vrf false` into account. I did not yet got the time to look at it in more details.

Thanks is advance.

**Comment by linouxis9** [comment]
Hi @abousselmi,

Yes, of course. With something like: `sudo ./packetrusher multi-ue -n 2 --tunnel --dedicatedGnb`.
I tried up to like 10 UEs, others users were successful as well.
Make sure your second gNodeB works (eg. you can ping its IP from outside).

Could you share some pcap?

Thanks and cheers,
Valentin

**Reply by linouxis9** [reply]
Hi @abousselmi,

Were you able to make it work?

Cheers,
Valentin

**Reply by abousselmi** [reply]
Hi @linouxis9 

Thanks for you reply. I was able to successfully test multiple UEs with a free5gc lab. I still can't make it work on another lab (based on a proprietary solution) but I found out that the issue is more related to the SMF behaviour, for some reason it deletes the PDU context of the second UE. So I'm trying to understand why. It might be related to PacketRusher but I cannot tell at this stage.

I let you know in any case. Thanks again !

Cheers,

**Reply by linouxis9** [reply]
No problem @abousselmi :-) Make sure the other SMF does not think that the PDU Session IDs are unique. All UEs can have the same PDU Session IDs!

Cheers,
Valentin

**Reply by abousselmi** [reply]
Thank @linouxis9 for the advise 👍 What I've noticed is that when the SMF informs the AMF about the creation of the context, it expects an answer (200 OK), which is sent by the AMF in the case of one UE and not the other.. To be continued..

Cheers,

**Reply by linouxis9** [reply]
> Hi @linouxis9
> 
> Thanks for you reply. I was able to successfully test multiple UEs with a free5gc lab. I still can't make it work on another lab (based on a proprietary solution) but I found out that the issue is more related to the SMF behaviour, for some reason it deletes the PDU context of the second UE. So I'm trying to understand why. It might be related to PacketRusher but I cannot tell at this stage.
> 
> I let you know in any case. Thanks again !
> 
> Cheers,

I'm wondering if your issue is related to https://github.com/HewlettPackard/PacketRusher/pull/135, you may try to see if this PR fixes your issue ;)

**Reply by abousselmi** [reply]
Hi @linouxis9 

Thanks a lot, I'll test it asap and let you know. 

Cheers!

**Reply by alekodu** [reply]
Hello, I am experiencing similar problems when trying to run a multi-ue test, it does work for multi-ue -n 1 when I add the tunnel and dedicatedGnb flags but it does not work when specifying more than one UE. I get the following error:

> │ + ./packetrusher --config config.yml multi-ue -n 2 --tunnel --dedicatedGnb                                                                                                                                     │
│ time="2024-09-19T13:17:09Z" level=info msg="Selecting 10.96.109.7 for host 10.96.109.7 as AMF's IP address"                                                                                                    │
│ time="2024-09-19T13:17:09Z" level=info msg="Selecting 10.244.0.29 for host 10.244.0.29 as gNodeB's N3/Data IP address"                                                                                         │
│ time="2024-09-19T13:17:09Z" level=info msg="Selecting 10.244.0.29 for host 10.244.0.29 as gNodeB's N2/Control IP address"                                                                                      │
│ time="2024-09-19T13:17:09Z" level=info msg="Loaded config at: config.yml"                                                                                                                                      │
│ time="2024-09-19T13:17:09Z" level=info msg="PacketRusher version 1.0.1"                                                                                                                                        │
│ time="2024-09-19T13:17:09Z" level=info msg=---------------------------------------                                                                                                                             │
│ time="2024-09-19T13:17:09Z" level=info msg="[TESTER] Starting test function: Testing registration of multiple UEs"                                                                                             │
│ time="2024-09-19T13:17:09Z" level=info msg="[TESTER][UE] Number of UEs: 2"                                                                                                                                     │
│ time="2024-09-19T13:17:09Z" level=info msg="[TESTER][GNB] gNodeB control interface IP/Port: 10.244.0.29/9487"                                                                                                  │
│ time="2024-09-19T13:17:09Z" level=info msg="[TESTER][GNB] gNodeB data interface IP/Port: 10.244.0.29/2152"                                                                                                     │
│ time="2024-09-19T13:17:09Z" level=info msg="[TESTER][AMF] AMF IP/Port: 10.96.109.7/38412"                                                                                                                      │
│ time="2024-09-19T13:17:09Z" level=info msg=---------------------------------------                                                                                                                             │
│ time="2024-09-19T13:17:09Z" level=info msg="[GNB] SCTP/NGAP service is running"                                                                                                                                │
│ time="2024-09-19T13:17:09Z" level=info msg="[GNB] Initiating NG Setup Request"                                                                                                                                 │
│ time="2024-09-19T13:17:09Z" level=fatal msg="Error incannot assign requested address"                                                                                                                          │
│ Stream closed EOF for default/packetrusher-6946db47b7-8g8kq (packetrusher)  

Could you please help me to figure out what I am doing wrong?

**Reply by linouxis9** [reply]
Hi @alekodu!
Do you have both 10.244.0.29 and 10.244.0.30 configured on your N2/N3 interface on your VM?
If it's a container, you may have to use multus to configure multiple IP address to your container.

**Reply by alekodu** [reply]
Thanks for the fast reply! I am running on a docker container on top of a Kind Kubernetes cluster. I may indeed need Multus in this case. I'll try it out...

**Reply by linouxis9** [reply]
> Thanks for the fast reply! I am running on a docker container on top of a Kind Kubernetes cluster. I may indeed need Multus in this case. I'll try it out...


If you are not planning to spawn millions of UEe, you can always run one container per UE with each having its own gNodeB -> it's the best for dataplane performance testing :-)

**Reply by alekodu** [reply]
So I thought as well and it works if I create a separate k8s deployment per UE-gNB but it does not work if I use the same deployment and scale to a larger number of replicas. With the scale up option, communication between UEs-gNBs is stopped and the AMF pod is restarted:

Latest replica:
![Screenshot 2024-09-20 at 10 58 49](https://github.com/user-attachments/assets/6420ac25-8e8e-4301-8c63-f7c740e417d8)

Older deployment pods:
![Screenshot 2024-09-20 at 11 04 50](https://github.com/user-attachments/assets/b80f1b96-f570-42ea-af6e-9a73032f36aa)

It would be more friendly if the user could just scale up the replicas. Do you know why this happens?

**Reply by linouxis9** [reply]
Make sure you are using a different gnbid for each gNodeB and a distinct MSIN for each UE. If you do not do that, the first UE to be registered will be released by the Core Network when the second UE with the same ID (IMSI/MSIN) will be released/freed by the Core Network!
I guess reusing the same deployment make each of the created pod using the same ConfigMap and as such the same MSIN and gnbId!

**Reply by alekodu** [reply]
You are right! Thanks for the explanation... Very useful tool you are building!

**Reply by linouxis9** [reply]
> You are right! Thanks for the explanation... Very useful tool you are building!

No problem! Happy to help 😁

**Reply by alekodu** [reply]
Hi again,

I am now testing packetrusher directly on a machine (instead of on docker). However, I keep having troubles when testing with more than one UE (enabling tunnelling and dedicatedGnb). You mentioned I need to configure consecutive IPs on the N2/N3 interface of my machine. I have done that by applying "sudo ip addr add ip brd broadcastip dev ethdev". Despite that, I do not manage to successfully test more than one UE. When I test with My5G-RANTester instead, it does create all tunnelling interfaces. Am I missing some configuration step for packetrusher to work?

**Reply by linouxis9** [reply]
Hi @alekodu!
Can you share the logs? Can you be more specific in what you have troubles with?
Are you able to ping each of the gNodeB IP from the Open5GS' VM for instance?
What does the PacketRusher's logs show?
Be aware that you won't see the interfaces if you do `ip a` as each UE interface will be in a distinct VRF, and you need to be in the VRF (ex. ip exec vrf000.. bash) to see the given interface.

Thanks and cheers,
Valentin
